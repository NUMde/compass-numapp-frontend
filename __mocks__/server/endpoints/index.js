import login from './login';
import getUser from './getUser';
import getLanguages from './getLanguages';
import updateLanguage from './updateLanguage';
import updateToken from './updateToken';
import getQuestionnaire from './getQuestionnaire';
import sendQuestionnaire from './sendQuestionnaire';
import report from './report';

export default [
  login,
  getUser,
  getLanguages,
  updateLanguage,
  updateToken,
  getQuestionnaire,
  sendQuestionnaire,
  report,
];
