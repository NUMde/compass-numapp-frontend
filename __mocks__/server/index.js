export { default as server } from './server';
