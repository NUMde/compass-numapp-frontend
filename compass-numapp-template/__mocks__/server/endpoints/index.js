import login from './login';
import getLanguages from './getLanguages';
import updateLanguage from './updateLanguage';
import updateToken from './updateToken';
import getQuestionnaire from './getQuestionnaire';
import sendQuestionnaire from './sendQuestionnaire';

export default [
  login,
  getLanguages,
  updateLanguage,
  updateToken,
  getQuestionnaire,
  sendQuestionnaire,
];
