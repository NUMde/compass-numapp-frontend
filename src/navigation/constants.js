/**
 * enum for all route names
 */
const Routes = {
  LANDING: 'Landing',
  LOGIN: 'Login',
  CHECK_IN: 'CheckIn',
  SURVEY: 'Survey',
  ABOUT: 'About',
  WEBVIEW: 'Webview',
  LEGAL_INFORMATION: 'LegalInformation',
};

/**
 * enum for all stacks
 */
const Stacks = {
  SIGNED_IN: 'SignedIn',
  SIGNED_OUT: 'SignedOut',
};

export { Routes, Stacks };
