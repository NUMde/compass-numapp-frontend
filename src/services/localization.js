// (C) Copyright IBM Deutschland GmbH 2021.  All rights reserved.

// this file provides functionality for adding additonal languages to the project.

/***********************************************************************************************
imports
***********************************************************************************************/

import i18n from 'i18n-js';
import memoize from 'lodash.memoize';
import { I18nManager } from 'react-native';
import * as RNLocalize from 'react-native-localize';

import de from '../CUSTOMIZATION/translations/de';
import en from '../CUSTOMIZATION/translations/en';
import fr from '../CUSTOMIZATION/translations/fr';
import ar from '../CUSTOMIZATION/translations/ar';

/***********************************************************************************************
constants
***********************************************************************************************/

// the fallback
// this is the default the app will fallback to in case nothing else matches.
// (for example: the device language is hindi, but only "de", "en", "fr", "ar" are provided by the app. So the last resort fallback would be "de")
export const defaultLanguage = 'de';

// the fallback
// this is the default reading direction the app will fallback to in case nothing else matches.
const defaultIsRTL = false;

// the available files / languages:
// for each file available in app/src/CUSTOMIZATION/translations an entry must be created here.
// the attribute name shall be the language code the file represents.
// if a language is present that is not provided with a matching questionnaire fom the backend, the language will remain available -
// the user will then be presented with the default questionnaire language.
const availableLanguageFiles = {
  de: {
    file: de,
    isRTL: false,
    title: 'Deutsch',
  },

  en: {
    file: en,
    isRTL: false,
    title: 'English',
  },

  fr: {
    file: fr,
    isRTL: false,
    title: 'Français',
  },

  ar: {
    file: ar,
    isRTL: true,
    title: 'عربي',
  },
};

// is used as the final composit of available languages
export const availableLanguages = {
  ...availableLanguageFiles,
};

// function that memorizes the results of i18n.t. Also determines the cache key for storing the result
const translate = memoize(
  (key, config) => i18n.t(key, config),
  (key, config) => (config ? key + JSON.stringify(config) : key),
);

// creates the set of languages available for the user
export const setAvailableLanguages = (list) => {
  // this function generates the union set of locally available languages and the languages the backend can provide the questionnaires in.

  const defaultLang = availableLanguageFiles[defaultLanguage];

  // adds new missing ones from the server
  list.data.forEach((langCode) => {
    if (!availableLanguages[langCode]) {
      availableLanguages[langCode] = {
        file: en,
        isRTL: defaultIsRTL,
        title: defaultLang
          ? `${defaultLang.title} (${
              defaultLang.file.generic.questionnaire
            }: ${langCode.replace(/^\w/, (c) => c.toUpperCase())})`
          : `DEFAULT (${langCode})`,
      };
    }
  });

  // updates the ones that do not have a matching entry from the backend
  Object.keys(availableLanguageFiles).forEach((key) => {
    if (!list.data.includes(key)) {
      availableLanguages[key].title += ` (${
        availableLanguages[key].file.generic.questionnaire
      }: ${defaultLanguage.replace(/^\w/, (c) => c.toUpperCase())})`;
    }
  });
};

// generates default values
const generateDefaultI18nConfigValues = () => {
  const fallback = { languageTag: defaultLanguage, isRTL: defaultIsRTL };
  // returns the language settings best matching the users device
  return (
    RNLocalize.findBestAvailableLanguage(Object.keys(availableLanguages)) ||
    fallback
  );
};

// sets the config
export const setI18nConfig = async (forcedLanguageTag) => {
  let finalLanguageTag = defaultLanguage;
  let finalIsFinalRTL;
  // generates the final config
  if (!forcedLanguageTag) {
    const { languageTag, isRTL } = generateDefaultI18nConfigValues();
    finalLanguageTag = languageTag;
    finalIsFinalRTL = availableLanguages[languageTag]
      ? availableLanguages[languageTag].isRTL
      : isRTL;
  } else {
    if (availableLanguages[forcedLanguageTag]) {
      finalLanguageTag = forcedLanguageTag;
    }
    finalIsFinalRTL = availableLanguages[forcedLanguageTag].isRTL;
  }
  // clear translation cache
  translate.cache.clear();
  // update layout direction
  I18nManager.forceRTL(finalIsFinalRTL);
  // set i18n-js config
  i18n.translations = {
    [finalLanguageTag]: availableLanguages[finalLanguageTag].file,
  };
  // finally setting the locale
  i18n.locale = finalLanguageTag;
};

// just returns the language tag
export const getLanguageTag = () => i18n.locale;

// initialization of the module - executed in App.jsx
export const initLocalization = (forcedLanguageTag) => {
  setI18nConfig(forcedLanguageTag);
};

/***********************************************************************************************
export
***********************************************************************************************/

export default translate;
