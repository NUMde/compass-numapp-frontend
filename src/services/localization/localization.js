// (C) Copyright IBM Deutschland GmbH 2021.  All rights reserved.

// this file provides functionality

/***********************************************************************************************
imports
***********************************************************************************************/

import i18n from "i18n-js";
import memoize from "lodash.memoize";
import { I18nManager } from "react-native";
import * as RNLocalize from "react-native-localize";

import de from '../../CUSTOMIZATION/translations/de'
import en from '../../CUSTOMIZATION/translations/en'
import fr from '../../CUSTOMIZATION/translations/fr'
import ar from '../../CUSTOMIZATION/translations/ar'

/***********************************************************************************************
constants
***********************************************************************************************/

const defaultLanguage = 'en';

const availableLanguageFiles = { de, en, fr, ar };  

const translate = memoize(
    (key, config) => i18n.t(key, config),
    (key, config) => (config ? key + JSON.stringify(config) : key)
);

const generateDefaultI18nConfigValues = () => {
    const fallback = { languageTag: defaultLanguage, isRTL: false };
    return RNLocalize.findBestAvailableLanguage(Object.keys(availableLanguageFiles)) || fallback;
};

const setI18nConfig = (forcedLanguageTag, isFinalRTL=false) => {
    let finalLanguageTag;

    if(!forcedLanguageTag) {
        const { languageTag, isRTL } = generateDefaultI18nConfigValues();
        finalLanguageTag = languageTag;
        isFinalRTL = isRTL;
    }
    else {
        finalLanguageTag = (availableLanguageFiles[forcedLanguageTag]) ? forcedLanguageTag : defaultLanguage;
    }

    // clear translation cache
    translate.cache.clear();
    // update layout direction
    I18nManager.forceRTL(isFinalRTL);
    // set i18n-js config
    i18n.translations = { [finalLanguageTag]: availableLanguageFiles[finalLanguageTag] };
    // finally setting the locale
    i18n.locale = finalLanguageTag;

    console.log('---> ' + finalLanguageTag)
};

const init = (forcedLanguageTag, isFinalRTL=false) => {
    setI18nConfig(forcedLanguageTag, isFinalRTL);
};

/***********************************************************************************************
export
***********************************************************************************************/

export default { translate, setI18nConfig, generateDefaultI18nConfigValues, defaultLanguage, init };
