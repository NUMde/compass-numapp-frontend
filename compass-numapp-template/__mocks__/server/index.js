import { setupServer } from 'msw/node';
import endpoints from './endpoints';

export default setupServer(...endpoints);
