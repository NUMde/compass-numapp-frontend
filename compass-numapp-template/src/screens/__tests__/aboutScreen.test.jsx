import React from 'react';
import { Alert, Linking } from 'react-native';

import {
  act,
  waitFor,
  fireEvent,
  renderWithRedux,
  waitForElementToBeRemoved,
} from '__test-utils__/render';
import server from '__mocks__/server';

import en from '~CUSTOMIZATION/translations/en';

import { getLanguageTag } from '~services/localization';

import AboutScreen from '~screens/aboutScreen';
import { Stacks, Routes } from '~navigation/constants';

describe('AboutScreen', () => {
  beforeAll(() => {
    server.listen();
  });

  afterEach(() => {
    server.resetHandlers();
    jest.clearAllMocks();
  });

  afterAll(() => {
    server.close();
  });

  it('should render the about screen', () => {
    const navigate = jest.fn();
    const goBack = jest.fn();
    const { toJSON } = renderWithRedux(
      <AboutScreen navigation={{ navigate, goBack }} />,
      { initialState: { User: { subjectId: 'someSubjectId' } } },
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should redirect to landing screen when reset was successful', async () => {
    const navigate = jest.fn();
    const goBack = jest.fn();
    const spyAlert = jest
      .spyOn(Alert, 'alert')
      .mockImplementation(async (title, message, callbackOrButtons) => {
        if (callbackOrButtons && callbackOrButtons[0]) {
          callbackOrButtons[0].onPress();
        }
      });
    const { getByText } = renderWithRedux(
      <AboutScreen navigation={{ navigate, goBack }} />,
      {
        initialState: {
          Globals: { loading: false },
          User: { subjectId: 'someSubjectId' },
        },
      },
    );
    const button = getByText(en.about.delete);
    expect(button).toBeTruthy();
    fireEvent.press(button);
    await waitFor(() =>
      expect(navigate).toHaveBeenCalledWith(Stacks.SIGNED_OUT, {
        screen: Routes.LANDING,
      }),
    );
    expect(spyAlert).toHaveBeenCalledWith(
      en.generic.warning,
      en.generic.eraseAllWarning,
      expect.anything(),
      expect.anything(),
    );
  });

  it('should change language', async () => {
    const navigate = jest.fn();
    const goBack = jest.fn();
    jest
      .spyOn(Alert, 'alert')
      .mockImplementation(async (title, message, callbackOrButtons) => {
        if (callbackOrButtons && callbackOrButtons[0]) {
          callbackOrButtons[0].onPress();
        }
      });
    const { getByTestId } = renderWithRedux(
      <AboutScreen navigation={{ navigate, goBack }} />,
      {
        initialState: {
          Globals: { loading: false },
          User: { subjectId: 'someSubjectId' },
        },
      },
    );
    const picker = getByTestId('languagePicker');
    act(() => picker.props.onChange({ nativeEvent: { newValue: 'de' } }));
    await waitForElementToBeRemoved(() => getByTestId('Spinner'));
    expect(getLanguageTag()).toBe('de');
  });

  it('should warn user before changing language when questionnaire has been started', async () => {
    const navigate = jest.fn();
    const goBack = jest.fn();
    const spyAlert = jest
      .spyOn(Alert, 'alert')
      .mockImplementation(async (title, message, callbackOrButtons) => {
        if (callbackOrButtons && callbackOrButtons[0]) {
          callbackOrButtons[0].onPress();
        }
      });
    const { getByTestId } = renderWithRedux(
      <AboutScreen navigation={{ navigate, goBack }} />,
      {
        initialState: {
          Globals: { loading: false },
          User: { subjectId: 'someSubjectId' },
          Questionnaire: {
            started: true,
          },
        },
      },
    );
    const picker = getByTestId('languagePicker');
    act(() => picker.props.onChange({ nativeEvent: { newValue: 'de' } }));
    await waitForElementToBeRemoved(() => getByTestId('Spinner'));
    expect(getLanguageTag()).toBe('de');
    expect(spyAlert).toHaveBeenCalledWith(
      en.generic.warning,
      en.about.languageWarning + en.about.languageWarningAddition,
      expect.anything(),
      expect.anything(),
    );
  });

  it('should show and hide redirectModal', async () => {
    const navigate = jest.fn();
    const goBack = jest.fn();
    const spyLinking = jest.spyOn(Linking, 'openURL');
    const { getByTestId, getByText } = renderWithRedux(
      <AboutScreen navigation={{ navigate, goBack }} />,
    );
    const modalLink = en.modalLinks[0];
    const item = getByText(modalLink.title);
    fireEvent.press(item);

    const button = getByTestId('redirectBtn');
    expect(button).toBeTruthy();
    fireEvent.press(button);

    expect(spyLinking).toHaveBeenCalledWith(modalLink.uri);
    await waitFor(() =>
      expect(getByTestId('redirectModal').props.visible).toBe(false),
    );
  });
});
