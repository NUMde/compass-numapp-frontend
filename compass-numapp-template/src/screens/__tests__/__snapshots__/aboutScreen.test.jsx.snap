// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AboutScreen should render the about screen 1`] = `
<Modal
  hardwareAccelerated={false}
  testID="Spinner"
  visible={true}
>
  <View
    style={
      {
        "backgroundColor": "#1F2532",
        "bottom": 0,
        "flex": 1,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <ActivityIndicator
        color="#EB576A"
        size="large"
        style={
          {
            "flex": 1,
          }
        }
      />
    </View>
  </View>
</Modal>
`;
